driver: can: add new filter to match CAN-FD frames

congnguyenhuu opened the pull request:
Add support FD frame filter to configure type frame for each Rx msg to receive corresponding frames (classic, FD frames). 

I have a PR https://github.com/zephyrproject-rtos/zephyr/pull/52450 which introduces support for NXP S32 CANEXCEL, it need to implement this.

congnguyenhuu committed:
driver: can: add new filter to match CAN-FD frames

Add support FD frame filter to configure type frame for
each Rx msg to receive corresponding frames (classic, FD frames).

The Bosch M_CAN driver does not support FD frame filter,
so inmplement driver to handle it in software.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
tests: drivers: can: canfd: add support filter type frame

Add new filter to configure correctly FD frame.

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

henrikbrixandersen reviewed (changes requested):
Still evaluating the idea of this API change, but if we are to add it we need the existing in-tree CAN-FD driver (yes, there's only one: Bosch M_CAN) adapted to support this and a test case added to prove it.

martinjaeger commented on include/zephyr/drivers/can.h:
```diff
@@ -212,7 +215,7 @@ struct can_filter {
 	uint32_t mask         : 29;
 	/** Flags. @see @ref CAN_FILTER_FLAGS. */
-	uint8_t flags         : 3;
+	uint8_t flags;
```
It's a bit unfortunate that we will increase the struct size by another word size for just this one bit. But on the other hand the switch to flags was exactly done for the purpose of being more extensible in the future, as you mentioned in #51361.

I was wondering if the flag should be enabled only for `CONFIG_CAN_FD_MODE=y` to lower the footprint for other use cases? Or would that introduce too much noise for little benefit?

henrikbrixandersen replied:
No, the flags should always be defined regardless of Kconfig settings. We were bound to hit a point where more than three bits were needed for the filter flags at some point, so I don't see any issue with this minor increase in RAM footprint.

congnguyenhuu commented:
> Member



> Still evaluating the idea of this API change, but if we are to add it we need the existing in-tree CAN-FD driver (yes, there's only one: Bosch M_CAN) adapted to support this and a test case added to prove it.

I think we should only add the filter type frame  to can drivers which require eg. NXP S32 CANEXCEL. I see Bosch M_CAN doesn't require this.
So I added a Kconfig to enable filter type frame. We can use if driver require. And I modified the test `canfd` to support this.
But now I don't have any board (run CAN-FD over Bosch M_CAN) to check it. Anyone have board, can help me run it?

henrikbrixandersen commented:
> I think we should only add the filter type frame to can drivers which require eg. NXP S32 CANEXCEL.

No. That will lead to different APIs for different CAN controllers, making it difficult to write portable application code. This is not an acceptable solution. 

> I see Bosch M_CAN doesn't require this.

It doesn't require it, but it could support it.

The above is exactly why I still hesitate in accepting this API change. We need to think carefully about user applications, portability, and support from more than just one driver/hardware peripheral.

congnguyenhuu commented:
> No. That will lead to different APIs for different CAN controllers, making it difficult to write portable application code. This is not an acceptable solution.

But in Bosch M_CAN driver, mcan filter no has field to configure filter type frame.
```
struct can_mcan_std_filter {
	volatile uint32_t id2  : 11; /* ID2 for dual or range, mask otherwise */
	volatile uint32_t res  :  5;
	volatile uint32_t id1  : 11;
	volatile uint32_t sfce :  3; /* Filter config */
	volatile uint32_t sft  :  2; /* Filter type */
} __packed __aligned(4);
```

henrikbrixandersen commented:
> But in Bosch M_CAN driver, mcan filter no has field to configure filter type frame.

It doesn't have a field for RTR either, but that is handled in software. Something similar can be done for FDF.

congnguyenhuu commented:
> It doesn't have a field for RTR either, but that is handled in software. Something similar can be done for FDF.

Great ! I handled in software as you said.

henrikbrixandersen reviewed (changes requested):
You need to update `can_utils_filter_match()` in `drivers/can/can_utils.h` to handle the new flag as well.

congnguyenhuu commented:
> You need to update `can_utils_filter_match()` in `drivers/can/can_utils.h` to handle the new flag as well.

I updated for handling it

henrikbrixandersen reviewed (dismissed):
Thanks! Looks good. Apologies for the review delay.

martinjaeger commented on drivers/can/can_mcan.h:
```diff
@@ -178,8 +178,10 @@ struct can_mcan_data {
+	uint32_t std_filt_fd_frame;
 	uint32_t std_filt_rtr;
 	uint32_t std_filt_rtr_mask;
+	uint8_t ext_filt_fd_frame;
```
I stumbled upon the difference in size for `std` and `ext` filters and found there is a potential variable overflow bug, see #53417. This should be `uint16_t`.

I have created a separate PR to fix the other two flags: #53418

congnguyenhuu replied:
> I stumbled upon the difference in size for `std` and `ext` filters and found there is a potential variable overflow bug, see #53417. This should be `uint16_t`.
> 
> I have created a separate PR to fix the other two flags: #53418

@henrikbrixandersen @martinjaeger  I updated size for `ext_filt_fd_frame`

congnguyenhuu replied:
Could you pls help me review this PR https://github.com/zephyrproject-rtos/zephyr/pull/52450.

martinjaeger reviewed (approved):
Looks good. Thanks for the fix.
